### PR TITLE
Add result_and_then_ok lint

### PR DESCRIPTION
**Pull request by printfn**
changelog: add `result_and_then_ok` lint to warn about `result.and_then(|o| Ok(x))` and suggest replacing with `result.map(|o| x)`.

### CODE REVIEW

File: clippy_lints/src/methods/mod.rs [hunk 1]
```diff
@@ -2191,6 +2221,97 @@ fn lint_option_and_then_some(cx: &LateContext<'_, '_>, expr: &hir::Expr, args: &
...
+    if !match_type(cx, ty, &paths::RESULT) {
+        return;
+    }
+
+    match args[1].node {
+        hir::ExprKind::Closure(_, _, body_id, closure_args_span, _) => {
+            let closure_body = cx.tcx.hir().body(body_id);
+            let closure_expr = remove_blocks(&closure_body.value);
+            if_chain! {
+                if let hir::ExprKind::Call(ref some_expr, ref some_args) = closure_expr.node;
```
Comment (lzutao) [inline clippy_lints/src/methods/mod.rs]:
```suggestion
                if let hir::ExprKind::Call(ref ok_expr, ref ok_args) = closure_expr.node;
```

File: clippy_lints/src/methods/mod.rs [hunk 2]
```diff
@@ -2191,6 +2221,97 @@ fn lint_option_and_then_some(cx: &LateContext<'_, '_>, expr: &hir::Expr, args: &
...
+                then {
+                    let inner_expr = &some_args[0];
+
+                    let mut finder = RetCallFinder { found: false };
+                    finder.visit_expr(inner_expr);
+                    if finder.found {
+                        return;
+                    }
+
+                    let some_inner_snip = if in_macro_or_desugar(inner_expr.span) {
```
Comment (lzutao) [inline clippy_lints/src/methods/mod.rs]:
```suggestion
                    let ok_inner_snip = if in_macro_or_desugar(inner_expr.span) {
```

File: tests/ui/option_and_then_some.rs [hunk 3]
```diff
@@ -9,10 +9,6 @@ pub fn main() {
     let _ = x.and_then(|o| Some(o + 1));
     // and an easy counter-example
     let _ = x.and_then(|o| if o < 32 { Some(o) } else { None });
-
-    // Different type
-    let x: Result<u32, &str> = Ok(1);
-    let _ = x.and_then(Ok);
```
Comment (lzutao) [inline tests/ui/option_and_then_some.rs]: Could you give a reason to remove these tests (for example compile errors)?
These tests are as regression tests for option_and_then_some lint.

Comment (printfn) [reply tests/ui/option_and_then_some.rs]: The new result_and_then_ok lint changes this to

```
    // Different type
    let x: Result<u32, &str> = Ok(1);
    let _ = x;
```

I could add it back in, but how could I then disable the result_and_then_ok lint from changing this file?

Comment (lzutao) [reply tests/ui/option_and_then_some.rs]: Add `#![deny(clippy::result_and_then_ok)]` to file I think.

Comment (printfn) [reply tests/ui/option_and_then_some.rs]: Will do

Comment (printfn) [reply tests/ui/option_and_then_some.rs]: It turned out I needed to add `#![allow(clippy::result_and_then_ok)]` to prevent Clippy from applying the lint.

**Comment by printfn** [comment]
Thanks for reviewing! I've now re-added the regression tests (with an `allow` attribute), renamed the variables and hopefully fixed the documentation code.

**Comment by flip1995** [comment]
Thanks for implementing this lint so fast after the other was finished!

You should be able to share most of the Code of the `option_and_then_some` lint. Please deduplicate as much code as possible and implement the two lints in one function. (That's actually the tricky part of adding this lint :wink:)

**Comment by hellow554** [comment]
@printfn please edit your first post and add one of the [closing keywords](https://help.github.com/en/articles/closing-issues-using-keywords) github provides together with the issue number 4412, e.g.

`closes #4412`

**Comment by flip1995** [comment]
implemented in #5529

**Pull request closed without merging**